File: typescript/src/policy.ts
/**
 * Client-side spending policy (mirrors the server session policy for fast,
 * local rejection before any network call). The server enforces again — this is
 * defence in depth, not the only gate.
 */
export interface Policy {
  perTxUsd?: number;
  dailyUsd?: number;
  allowedTokens?: string[];     // e.g. ["USDC","WETH"] — empty/undefined = any
  allowedRecipients?: string[]; // 0x… lowercased — empty/undefined = any
}

let spentToday = 0;
let spentDay = '';

export function localPolicyCheck(p: Policy, tx: { token: string; recipient: string; estUsd: number }): { ok: boolean; reason?: string } {
  if (p.allowedTokens?.length && !p.allowedTokens.map((t) => t.toUpperCase()).includes(tx.token.toUpperCase())) {
    return { ok: false, reason: 'token not allowed' };
  }
  if (p.allowedRecipients?.length && !p.allowedRecipients.map((r) => r.toLowerCase()).includes(tx.recipient.toLowerCase())) {
    return { ok: false, reason: 'recipient not allowed' };
  }
  if (p.perTxUsd && tx.estUsd > p.perTxUsd) return { ok: false, reason: 'over per-tx cap' };
  if (p.dailyUsd) {
    const day = new Date().toISOString().slice(0, 10);
    if (day !== spentDay) { spentDay = day; spentToday = 0; }
    if (spentToday + tx.estUsd > p.dailyUsd) return { ok: false, reason: 'over daily cap' };
    spentToday += tx.estUsd;
  }
  return { ok: true };
}

File: python/acechange_agent_wallet/policy.py
"""Client-side spending policy (mirrors the server session policy for fast local
rejection). The server enforces again — defence in depth, not the only gate."""
from __future__ import annotations
from dataclasses import dataclass, field
from datetime import date


@dataclass
class Policy:
    per_tx_usd: float = 0.0
    daily_usd: float = 0.0
    allowed_tokens: list[str] = field(default_factory=list)      # [] = any
    allowed_recipients: list[str] = field(default_factory=list)  # [] = any
    _spent: float = 0.0
    _day: str = ""

    def check(self, token: str, recipient: str, est_usd: float) -> tuple[bool, str]:
        if self.allowed_tokens and token.upper() not in [t.upper() for t in self.allowed_tokens]:
            return False, "token not allowed"
        if self.allowed_recipients and recipient.lower() not in [r.lower() for r in self.allowed_recipients]:
            return False, "recipient not allowed"
        if self.per_tx_usd and est_usd > self.per_tx_usd:
            return False, "over per-tx cap"
        if self.daily_usd:
            today = date.today().isoformat()
            if today != self._day:
                self._day, self._spent = today, 0.0
            if self._spent + est_usd > self.daily_usd:
                return False, "over daily cap"
            self._spent += est_usd
        return True, ""
